### PR TITLE
TagInput improvements (#1412)

### LINKED ISSUE #1412: TagInput improvements
Collecting a bunch of features I want to add to `TagInput`:

- [ ] 1. `inputValue`, `onInputChange` props to listen/control input without resorting to `inputProps`
- [x] 2. #1418 `placeholder` prop that is removed when `values.length > 0` (use `inputProps.placeholder` if you always want it there)
- [x] 3. `InputGroup` support for left icon / right element (needs styles) #1420
- [x] 4. #1418 `separator` prop automatically splits input value to add multiple values at once (default: `/,\s*/g`)
- [ ] 5. option to add values on `blur` (in addition to <kbd>enter</kbd>)
- [x] 6. #1418 `onChange` receives `string[]` and handles the add / remove array manipulation internally
- [x] 7. `disabled` support (needs styles) https://github.com/palantir/blueprint/pull/1819
- [x] 8. #1419 support for JSX `values`

### PR DESCRIPTION
**Pull request by kmblake**
#### Checklist

- [x] [Enable CircleCI for your fork](https://circleci.com/add-projects)
- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

inputValue, onInputChange props to listen/control input without resorting to inputProps
option to add values on blur (in addition to enter) via addTagOnBlur prop

### CODE REVIEW

File: packages/labs/src/components/tag-input/tagInput.tsx [hunk 1]
```diff
@@ -34,6 +41,9 @@ export interface ITagInputProps extends IProps {
     /** React props to pass to the `<input>` element */
     inputProps?: HTMLInputProps;
 
+    /** Initial value of the `<input>` element */
```
Comment (giladgray) [inline packages/labs/src/components/tag-input/tagInput.tsx]: controlled value, not initial.

File: packages/labs/src/components/tag-input/tagInput.tsx [hunk 2]
```diff
@@ -148,13 +161,14 @@ export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState>
...
         inputProps: {},
+        inputValue: "",
         separator: ",",
         tagProps: {},
     };
 
     public state: ITagInputState = {
         activeIndex: NONE,
-        inputValue: "",
+        inputValue: this.props.inputValue,
```
Comment (giladgray) [inline packages/labs/src/components/tag-input/tagInput.tsx]: hmm this is not the implementation i expected. this is a `defaultInputValue`, but what we really want is an actual controlled `inputValue`, which means that updating the prop in the future will be reflected in the state (not just on initial load like you have here).

File: packages/labs/src/components/tag-input/tagInput.tsx [hunk 3]
```diff
@@ -23,6 +23,13 @@ import {
 import * as Classes from "../../common/classes";
 
 export interface ITagInputProps extends IProps {
+    /**
+     * If this is set to true, a new tag will be created from a populated input field when
+     * the user either clicks out of the component or presses `enter` on the input.
+     * @default false
+     */
+    addTagOnBlur?: boolean;
```
Comment (giladgray) [inline packages/labs/src/components/tag-input/tagInput.tsx]: i think simply `addOnBlur` should be sufficient

File: packages/labs/src/components/tag-input/tagInput.tsx [hunk 4]
```diff
@@ -23,6 +23,13 @@ import {
 import * as Classes from "../../common/classes";
 
 export interface ITagInputProps extends IProps {
+    /**
+     * If this is set to true, a new tag will be created from a populated input field when
+     * the user either clicks out of the component or presses `enter` on the input.
```
Comment (giladgray) [inline packages/labs/src/components/tag-input/tagInput.tsx]: "If `true`, `onAdd` will also be invoked when the input loses focus. By default, `onAdd` is only invoked on `enter`."

Comment (kmblake) [reply packages/labs/src/components/tag-input/tagInput.tsx]: Addressed in https://github.com/palantir/blueprint/pull/1966

File: packages/labs/src/components/tag-input/tagInput.tsx (hunk 3 above)
Comment (kmblake) [reply packages/labs/src/components/tag-input/tagInput.tsx]: Addressed in https://github.com/palantir/blueprint/pull/1966

File: packages/labs/src/components/tag-input/tagInput.tsx (hunk 2 above)
Comment (kmblake) [reply packages/labs/src/components/tag-input/tagInput.tsx]: Addressed in https://github.com/palantir/blueprint/pull/1967

File: packages/labs/src/components/tag-input/tagInput.tsx (hunk 1 above)
Comment (kmblake) [reply packages/labs/src/components/tag-input/tagInput.tsx]: Addressed in https://github.com/palantir/blueprint/pull/1967

File: packages/labs/src/components/tag-input/tag-input.md [hunk 5]
```diff
@@ -10,13 +10,16 @@
...
 
-`TagInput` provides granular `onAdd` and `onRemove` **event props**, which are passed the added or removed items in response to the user interactions above. It also provides `onChange`, which combines both events and is passed the updated `values` array, with new items appended to the end and removed items filtered away. Supply `inputProps` to customize the `<input>` element or add your own event handlers.
+`TagInput` provides granular `onAdd` and `onRemove` **event props**, which are passed the added or removed items in response to the user interactions above. It also provides `onChange`, which combines both events and is passed the updated `values` array, with new items appended to the end and removed items filtered away.
+
+To set a default value for the `<input>` element, supply an `inputValue` prop. Use `onInputChange` to respond to changes to the `<input>` text. Additional properties (such as custom event handlers) can be applied to the `<input>` element via `inputProps`.
```
[the remark this answers is not in the document]
Comment (kmblake) [reply packages/labs/src/components/tag-input/tag-input.md]: Ah, got it. Fixed in https://github.com/palantir/blueprint/pull/1967

**Pull request closed without merging**